app/databases.py

```diff
@@ -0,0 +1,153 @@
+    """
+
+    client = CosmosClient(
+        os.environ["COSMOSDB_ENDPOINT"],
+        credential=(
+            DefaultAzureCredential()
+            if environment != "local"
+            else os.environ["COSMOSDB_KEY"]
+        ),
+        connection_verify=(environment != "local"),
+    )
+    logging.info("Cosmos client created.")
+    return client
+
+
+class CosmosDBLongCallbackManager:
+    '''
+    Use CosmosDB as a cache. Takes inputs of 
+        cosmos_client as above
+        database_name
+        container_name (created if does not exist)
+        expiry time (in seconds)
+        partition_key (optional; else "/Id").
+
+    the "get" method retrieves the cached value 
+    unless it is expired.
+
+    the "set" method inputs a value to the cache
+    '''
+    def __init__(
+        self,
+        cosmos_client: CosmosClient,
+        database_name: str,
+        container_name: str,
+        expire: int,
+        partition_key: str = "/Id",
+    ):
+        self.client = cosmos_client
+        self.database_name = database_name
+        self.container_name = container_name
+        self.expire = expire
+        self.partition_key = partition_key
+
+
+    def is_expired(self, item):
+        if not item or "timestamp" not in item or "expire" not in item:
+            return True
+        timestamp = datetime.fromisoformat(item["timestamp"])
+        expire_seconds = item["expire"]
+        now = datetime.now(timezone.utc)
+        seconds_since_timestamp = (now - timestamp).total_seconds()
+        if seconds_since_timestamp > expire_seconds:
+            logging.info(
+                f"Now:{now}. timestamp: {timestamp}. seconds_since_stimestamp: {seconds_since_timestamp}"
+            )
+            return True
+        return False
+
+    def get_database_client(self):
+        return self.client.get_database_client(self.database_name)
+
+    def get_container_client(self):
+        return self.get_database_client().get_container_client(self.container_name)
+
+    def get(self, cache_key):
+        container = self.get_container_client()
+        query = f'SELECT * FROM c WHERE c.id = "{cache_key}"'
+        items = list(container.query_items(query, enable_cross_partition_query=True))
+
+        if not items:
+            return None
+
+        item = items[0]
+
+        if self.is_expired(item):
+            return None
+
+        else:
+            return item.get("value")
+
+    def set(self, value, cache_key):
+        container = self.get_container_client()
+        if isinstance(value, datetime):
+            value = value.isoformat()
+        item = {
+            "id": cache_key,
+            "value": value,
+            "expire": self.expire,
+            "timestamp": datetime.now(timezone.utc).isoformat(),
+        }
+
+        if self.partition_key is not None:
+            item[self.partition_key] = self.partition_key
+
+        query = f'SELECT * FROM c WHERE c.id = "{cache_key}"'
```

Comment:
Are we protecting the cache from containing multiple values, or is this actually OK?
We're doing `select *` which could bring n rows back. If that happens, what errors or behaviour might we see?

Reply:
We check if an item exists with the specified cache_key (which is hard-coded at present), and update if it does exist, else insert a new item. 

If two users were to try to insert at the same time, we have no cache "lock" mechanism - but I suppose (after the very first refresh) all users would be "upsert"ing", so whoever hits the server second would wipe the first user's cache. 

In this case, since the cache is only updated through the hyacinth app, I think the content of the cache being stored by user 1 and user 2 would be identical - so it wouldn't break

Reply:
I think the key point here is that the cache key we use is hard-coded (or could be set by an environment variable - either way, not a random altering variable). You now have access to the hyacinth repo - an example of its use is in the callback [here](https://github.com/UCLH-FlowEHR-Bed/hyacinth/blob/main/app/pages/discharges/callbacks.py)
```python
DATABASE_KEY = "database_cache"

cosmosdb_callback_manager = CosmosDBLongCallbackManager(
        cosmos_client(),
        database_name="hyacinth-state",
        container_name="hyacinth",
        expire=(60*2),
        partition_key="/id",
    )
    use_cosmos = True

## when using cosmos:
        cached_data = cosmosdb_callback_manager.get(DATABASE_KEY)

## currently we set with a nasty pickle 
        cosmosdb_callback_manager.set(
            # Dump object as base64 encoded pickle file for data consistency
            cache_key=DATABASE_KEY, value=codecs.encode(pickle.dumps(df), "base64").decode()
    )

```

We pickle because the datetimes got messed up otherwise. For consistency we could set the cosmos callback manager to do this pickling automatically. With the demo tomorrow it seemed like the wrong thing to focus on though!